driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java

```diff
@@ -50,30 +70,90 @@
          * @param host the host this policy applies to
          * @return the newly created {@link ConvictionPolicy} instance.
          */
-        public ConvictionPolicy create(Host host);
+        ConvictionPolicy create(Host host, ReconnectionPolicy reconnectionPolicy);
     }
 
-    public static class Simple extends ConvictionPolicy {
-        private Simple(Host host) {
+    static class DefaultConvictionPolicy extends ConvictionPolicy {
+        private final Host host;
+        private final ReconnectionPolicy reconnectionPolicy;
+        private final AtomicInteger openConnections = new AtomicInteger();
+
+        private volatile long nextReconnectionTime = Long.MIN_VALUE;
+        private ReconnectionPolicy.ReconnectionSchedule reconnectionSchedule;
+
+        private DefaultConvictionPolicy(Host host, ReconnectionPolicy reconnectionPolicy) {
+            this.host = host;
+            this.reconnectionPolicy = reconnectionPolicy;
         }
 
         @Override
-        public boolean addFailure(ConnectionException exception) {
-            return true;
+        void signalConnectionCreated() {
+            int newCount = openConnections.incrementAndGet();
+            Host.statesLogger.debug("[{}] new connection created, total = {}", host, newCount);
+            resetReconnectionTime();
         }
 
-        public boolean addFailureFromExternalDetector() {
-            return true;
+        @Override
+        void signalConnectionClosed() {
+            if (host.state == Host.State.DOWN)
+                return;
+            int remaining = openConnections.decrementAndGet();
+            assert remaining >= 0;
```

Comment:
I think for this to happen you have a Connection with pending requests on a close sitting in the trash.  Noticed this in my logs slightly before:

```
    41927  [cluster2-scheduled-task-worker-0] DEBUG com.datastax.driver.core.Connection  - Not terminating Connection[/127.0.1.1:9042-2, inFlight=0, closed=true]: there are still pending requests
```

**Note**:  This was a bit of a bug in my code, I'm returning on a connection but not emptying the dispatcher when I return the connection, but it creates this scenario which I thought was interesting.

It triggers a heartbeat on that connection which is closed and in the trash:

```
    61944  [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection  - Connection[/127.0.1.1:9042-2, inFlight=0, closed=true] was inactive for 30 seconds, sending heartbeat
    61944  [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection  - Not terminating Connection[/127.0.1.1:9042-2, inFlight=0, closed=true]: there are still pending requests
    61944  [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection  - Connection[/127.0.1.1:9042-2, inFlight=0, closed=true] connection error
com.datastax.driver.core.ConnectionException: [/127.0.1.1:9042] Connection has been closed
	at com.datastax.driver.core.Connection.write(Connection.java:520)
	at com.datastax.driver.core.Connection.write(Connection.java:496)
	at com.datastax.driver.core.Connection$Dispatcher.userEventTriggered(Connection.java:1000)
	at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.timeout.IdleStateHandler.channelIdle(IdleStateHandler.java:322)
	at io.netty.handler.timeout.IdleStateHandler$AllIdleTimeoutTask.run(IdleStateHandler.java:435)
	at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:357)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	at java.lang.Thread.run(Thread.java:745)
```

This signals a failure on the connection (I think) which decrements the count.

If I add !isClosed to the condition check in userEventTriggered in Connection$Dispatcher the error goes away since we don't try sending a heartbeat on a closed connection in the trash.